### PR TITLE
Make `move_sprite` both more general and more illustrative.

### PR DESCRIPTION
**Pull request by stephenmartindale**
# Objective

Discussion: https://discord.com/channels/691052431525675048/743559241461399582/1063153669920526376

This example is important because it is far simpler than the breakout and ECS-guide examples and thus likely to be encountered by newcomers to *Bevy*. Furthermore, unlike the breakout example which uses a `1/60`-second time-step, this example demonstrates movement with truly continuous time.

The previous version of this code has some shortcomings. Primarily, the `Direction` component is an `enum` and is mutated by the `sprite_movement(…)` system, suggesting that its reason for existence is as the mutable state required to implement the up-and-down movement rules.

Its purpose as a *marker* component is subtle and that represents a missed opportunity as well as producing an unintended pitfall for newcomers who might copy this examples code and remove that component – perhaps implementing movement with periodic functions like `sin()`.

## Solution

The revised example implements identical behaviour but places the emphasis on two key lessons for the reader:

1. How one might use a system to smoothly update the translation of sprites, moving them over time.
2. That components are widely re-used – particularly, Transform – and so queries must be well defined and marker components have their place.

Additionally, the style has been made more appropriate for the domain. For example, the movement update is now just `translation += v × dt`.

### CODE REVIEW

File: examples/2d/move_sprite.rs [hunk 1]
```diff
@@ -1,46 +1,54 @@
...
-    for (mut logo, mut transform) in &mut sprite_position {
-        match *logo {
-            Direction::Up => transform.translation.y += 150. * time.delta_seconds(),
-            Direction::Down => transform.translation.y -= 150. * time.delta_seconds(),
-        }
+/// Update everything that has both a velocity *and* a transform.
+fn update_moveables(time: Res<Time>, mut query: Query<(&Velocity, &mut Transform)>) {
+    for (v, mut trf) in &mut query {
+        trf.translation += Vec3::from((v.0 * time.delta_seconds(), 0.));
+    }
```
Comment (nicopap) [inline examples/2d/move_sprite.rs:39]:
```suggestion
    let delta = time.delta_seconds();
    for (velocity, mut transform) in &mut query {
        transform.translation += velocity.0.extend(0.0) * delta;
    }
```

I think in this particular case the whole name is better. To improve the math line, you could add a binding for `time.delta_seconds()`.

File: examples/2d/move_sprite.rs [hunk 2]
```diff
@@ -1,46 +1,54 @@
...
-/// The sprite is animated by changing its translation depending on the time that has passed since
-/// the last frame.
-fn sprite_movement(time: Res<Time>, mut sprite_position: Query<(&mut Direction, &mut Transform)>) {
-    for (mut logo, mut transform) in &mut sprite_position {
-        match *logo {
-            Direction::Up => transform.translation.y += 150. * time.delta_seconds(),
-            Direction::Down => transform.translation.y -= 150. * time.delta_seconds(),
-        }
+/// Update everything that has both a velocity *and* a transform.
+fn update_moveables(time: Res<Time>, mut query: Query<(&Velocity, &mut Transform)>) {
```
Comment (nicopap) [inline examples/2d/move_sprite.rs:36]:
```suggestion
fn apply_velocity(time: Res<Time>, mut query: Query<(&Velocity, &mut Transform)>) {
```

`update_moveables` here doesn't describe what the system does. For one, I find `moveable` out of place since it's not used anywhere else.

File: examples/2d/move_sprite.rs [hunk 3]
```diff
@@ -1,46 +1,54 @@
...
+/// Update bird sprites according to our specific rules.
+///
+/// Many components are widely used and few are as ubiquitous as `Transform`;
+/// component reuse is one of the core tenets of Entity Component Systems (ECS).
+///
+/// This system uses a filter – `With<Bird>` – to restrict updates to bird sprites, only.
+fn update_birds(mut query: Query<(&mut Velocity, &Transform), With<Bird>>) {
+    for (mut v, trf) in &mut query {
+        if trf.translation.y.abs() > 200. {
+            v.0 *= -1.;
```
Comment (nicopap) [inline examples/2d/move_sprite.rs:51]: :+1: this makes a lot of sense, and that's actually a great starting point for someone who want to extend the example for their own use. Notwistanding, it should use full variable names like `velocity` and `transform`.

File: examples/2d/move_sprite.rs [hunk 4]
```diff
@@ -1,46 +1,54 @@
...
     App::new()
         .add_plugins(DefaultPlugins)
         .add_startup_system(setup)
-        .add_system(sprite_movement)
+        .add_system(update_moveables)
+        .add_system(update_birds)
         .run();
 }
 
+/// A component for any sprite that might move.
```
Comment (nicopap) [inline examples/2d/move_sprite.rs:14]:
```suggestion
/// A component for any entity that might move.
```

**Review by nicopap — CHANGES_REQUESTED** [review]
I think the change improves the example tenfold. Though I'd like to see the variable names be more full. Since `Velocity` and `Transform` are struct, I don't think it makes sense to give them "mathy" names. I'd only expect math primitives (like integers or `Vec3`) to have one-letter variables.

File: examples/2d/move_sprite.rs [hunk 5]
```diff
@@ -1,46 +1,54 @@
...
 
 use bevy::prelude::*;
 
 fn main() {
     App::new()
         .add_plugins(DefaultPlugins)
         .add_startup_system(setup)
-        .add_system(sprite_movement)
+        .add_system(update_moveables)
+        .add_system(update_birds)
```
Comment (nicopap) [inline examples/2d/move_sprite.rs:10]: The system order matters here. An inconsistent ordering will result in stuttering and surprising delay.

```suggestion
        .add_system(update_moveables)
        .add_system(update_birds.before(update_moveables))
```

File: examples/2d/move_sprite.rs [hunk 6]
```diff
@@ -1,46 +1,54 @@
...
         .run();
 }
 
+/// A component for any sprite that might move.
 #[derive(Component)]
-enum Direction {
-    Up,
-    Down,
-}
+struct Velocity(Vec2);
```
Comment (rparrett) [inline examples/2d/move_sprite.rs:16]: We are deriving `Deref` and `DerefMut` to clean up the `.0` spam in (some) other examples. Should we do the same here?

File: examples/2d/move_sprite.rs (hunk 5 above)
Comment (rparrett) [reply examples/2d/move_sprite.rs:10]: Perhaps also add a comment reflecting that?

Comment (stephenmartindale) [reply examples/2d/move_sprite.rs:10]: I thought about that but chose to leave out the ordering because I could not decide that tutorialising sytem dependencies was in-scope for the example. But, since you suggested the change, I now have an argument to do so.

File: examples/2d/move_sprite.rs (hunk 4 above)
Comment (stephenmartindale) [reply examples/2d/move_sprite.rs:14]: I like the noun "sprite" because it fits the branding of the example: move sprites.

How about "A component for anyTHING moving."?

File: examples/2d/move_sprite.rs (hunk 5 above)
Comment (rparrett) [reply examples/2d/move_sprite.rs]: I wouldn't necessarily "tutorialize system dependencies" but a brief comment explaining why this particular ordering is important would be nice.

File: examples/2d/move_sprite.rs (hunk 6 above)
Comment (stephenmartindale) [reply examples/2d/move_sprite.rs:16]: I also thought about that. It would add ~~10 lines of code (guess!) to a 50-line example, though. Would that really be worth it?

Comment (rparrett) [reply examples/2d/move_sprite.rs:16]: It wouldn't add any lines of code
```diff
diff --git a/examples/2d/move_sprite.rs b/examples/2d/move_sprite.rs
index 5290ac1e2..88b78f1bb 100644
--- a/examples/2d/move_sprite.rs
+++ b/examples/2d/move_sprite.rs
@@ -12,7 +12,7 @@ fn main() {
 }
 
 /// A component for any sprite that might move.
-#[derive(Component)]
+#[derive(Component, Deref, DerefMut)]
 struct Velocity(Vec2);
 
 /// A marker component for our particular sprite.
@@ -36,7 +36,7 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
 fn update_moveables(time: Res<Time>, mut query: Query<(&Velocity, &mut Transform)>) {
     let delta = time.delta_seconds();
     for (velocity, mut transform) in &mut query {
-        transform.translation += velocity.0.extend(0.0) * delta;
+        transform.translation += velocity.extend(0.0) * delta;
     }
 }
 
@@ -49,7 +49,7 @@ fn update_moveables(time: Res<Time>, mut query: Query<(&Velocity, &mut Transform
 fn update_birds(mut query: Query<(&mut Velocity, &Transform), With<Bird>>) {
     for (mut v, trf) in &mut query {
         if trf.translation.y.abs() > 200. {
-            v.0 *= -1.;
+            **v *= -1.;
         }
     }
 }
 ```

Comment (stephenmartindale) [reply examples/2d/move_sprite.rs:16]: Oh gosh. I never `#[derive…]` those in my own code-base so I totally forgot you *could*.

It's definitely worth doing, then.

Comment (stephenmartindale) [reply examples/2d/move_sprite.rs:16]: One other thing: You call `.extend(0.0)` – which matches my own style for `f32`s (`.0`) – but I had originally copied the original examples style (`0.`). Which one is preferred?
(If I had a say, I'd go with `.0` but then I would also patch all the other `f32` literals in the file.)

Comment (rparrett) [reply examples/2d/move_sprite.rs:16]: That's not my code, but I prefer `0.`.

Comment (rparrett) [reply examples/2d/move_sprite.rs]: At a glance it seems like we're generally doing `0.0`

Comment (stephenmartindale) [reply examples/2d/move_sprite.rs]: Ok. I added `Deref` and `DerefMut` as the latest commit to the branch but... the result isn't very satisfying.

I don't like the double-dereference-mul-asign idiom in your patch – that would be a show-stopper for any non-Rustacean reader in otherwise trivially easy code – so I left that out... but... yeah. I'm not sold on the end-product.

File: examples/2d/move_sprite.rs (hunk 2 above)
Comment (stephenmartindale) [reply examples/2d/move_sprite.rs]: I've renamed this system to `apply_velocity` and I renamed the other system to `update_bird_velocity`, then also re-ordered them so that the reader reads them in the order they actually run:

```rust
        .add_system(update_bird_velocity)
        .add_system(apply_velocity.after(update_bird_velocity))
```

I also patched the comment so that it is symmetric with the comment above the `Velocity` component.

I'll resolve this conversation but feel free to re-open.

File: examples/2d/move_sprite.rs (hunk 3 above)
Comment (stephenmartindale) [reply examples/2d/move_sprite.rs]: Full variable names used – they're my own preference, actually. There must have been something in my coffee or else I was still mentally stuck on the type-name: `Quat`... grr... `Quat`...

File: examples/2d/move_sprite.rs (hunk 4 above)
Comment (stephenmartindale) [reply examples/2d/move_sprite.rs]: I have changed this comment to be:

```rs
/// A component for any moving entity.
```

... and also that above `apply_velocity` so that they're symmetrical:

```rs
/// Update all moving entities: those with a velocity and a transform.
```

**Comment by mockersf** [comment]
I think introducing two systems and requiring ordering between them is too complex for this example and not worth it.

File: examples/2d/move_sprite.rs (hunk 4 above)
Comment (nicopap) [reply examples/2d/move_sprite.rs]: My first idea was to suggest `anything`, but then I found the phrasing awkward, so I preferred replacing with "entity". But the new comment is great.

**Review by nicopap — APPROVED** [review]
I'm split on this.

When I proposed this structure for the example, I said that "it was over-engineered". But, as you pointed out in discord, if the examples serve as starting point for using/learning bevy, they should reflect how a bevy app looks. And this _is_ pretty much how it would look like.

I understand that the current policy for examples is to do exactly one thing and be focused. Which helps use them as tests. It also has a pedagogical purpose, because it only introduces a single concept and enable the reader to focus on understanding only a single thing at a time, to not be overwhelmed.

But following this policy also prevents us from exposing the user to more principled bevy apps, that can be easily changed and evolve smoothly.

It also comes down to how people use bevy examples: Examples are pieces of functional code that you can just run, but also tweak to see your code change, without much more time investment. Ideally examples should nudge the new user toward practices that will help them smoothly build up an app. We can't really influence how people use examples, so I think making sure they are extensible is probably a good idea.

Overall, I think splitting the system in two and having the example follow the structure of an actual bevy app balances the cost of having to explicit the order of the systems. Also, is `foo.after(bar)` so much cryptic that it is a deal breaker?

tl;dr

Pros:
* Users will assume examples are good starting point for their own games, this allows user to evolve examples into their own game smoothly.
* There is no way to tell people to not use examples as starting point, so we might as make it smooth for them.

Cons:
* the `.after` system label may be cryptic and confuse people that just want to see how to move a sprite.
* Technically, this doesn't only introduce how to move a sprite, but also how to order systems.
* People may think bevy is overly verbose because you need to write "so much code" just to move a sprite on screen (+10 lines compared to original, though those are mostly comments).
